api/s2n.h

```diff
@@ -596,6 +596,210 @@ extern int s2n_cert_get_utf8_string_from_extension_data_length(const uint8_t *ex
+ * @param identity The PSK identity being set.
+ * @param identity_size The length of the PSK identity being set.
+ */
+S2N_API 
+int s2n_psk_set_identity(struct s2n_psk *psk, const uint8_t *identity, uint16_t identity_size);
+
+/**
+ * Sets the PSK secret for a given PSK object. 
+ * 
+ * @param psk A pointer to PSK object to be updated with the PSK secret.
+ * @param secret The PSK secret in hex-encoded format.
+ * @param secret_size The length of the PSK secret being set.
+ */
+S2N_API 
+int s2n_psk_set_secret(struct s2n_psk *psk, const uint8_t *secret, uint16_t secret_size);
+
+/**
+ * Sets the PSK HMAC algorithm for a given PSK object. 
+ * 
+ * @param psk A pointer to the PSK object to be updated with the PSK HMAC algorithm.
+ * @param hmac The PSK HMAC algorithm being set. 
+ * The supported PSK HMAC algorithms are listed in the enum `s2n_psk_hmac` above. 
+ */
+S2N_API 
+int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac);
+
+/**
+ * Appends a PSK to the s2n connection object. If a duplicate PSK identity is found, 
+ * S2N_ERR_DUPLICATE_PSK_IDENTITIES error is thrown.
+ * 
+ * @param conn A pointer to the s2n_connection object to be updated with the PSK.
+ * @param psk A pointer to the PSK object to be appended.
+ */
+S2N_API 
+int s2n_connection_append_psk(struct s2n_connection *conn, struct s2n_psk *psk);
+
+/* s2n PSK supported modes for TLS1.3 */ 
+typedef enum { S2N_PSK_MODE_RESUMPTION, S2N_PSK_MODE_EXTERNAL } s2n_psk_mode;
+
+/**
+ * Sets the PSK mode on the s2n config object. 
+ * 
+ * @param config A pointer to the s2n_config object being updated.
+ * @param mode The PSK mode to be set. The supported psk modes are listed in the enum `s2n_psk_mode` above. 
+ */
+S2N_API 
+int s2n_config_set_psk_mode(struct s2n_config *config, s2n_psk_mode mode);
+
+/**
+ * Sets the PSK mode on the s2n connection object. 
+ * 
+ * @param conn A pointer to the s2n_connection object being updated.
+ * @param mode The PSK mode to be set. The supported psk modes are listed in the enum `s2n_psk_mode` above. 
+ */
+S2N_API 
+int s2n_connection_set_psk_mode(struct s2n_connection *conn, s2n_psk_mode mode);
+
+/**
+ * Gets the negotiated PSK identity length from the s2n connection object. 
+ * 
+ * @param conn A pointer to the s2n_connection object that successfully negotiated a PSK connection.
+ * @param identity_length The length of the negotiated PSK identity. 
+ */
+S2N_API 
+int s2n_connection_get_negotiated_psk_identity_length(struct s2n_connection *conn, uint16_t *identity_length);
+
+/**
+ * Gets the negotiated PSK identity from the s2n connection object. 
+ * 
+ * @param conn A pointer to the s2n_connection object that successfully negotiated a PSK connection.
+ * @param identity The negotiated PSK identity obtained from the s2n_connection object. 
+ * @param max_identity_length The maximum length for the PSK identity. If the negotiated psk_identity length is 
+ * greater than this value `S2N_ERR_INSUFFICIENT_MEM_SIZE` error is thrown.
```

Comment:
Giving exact error codes might be too specific. Are exact error codes part of our contract? We haven't traditionally treated them as set in stone, although where or not we return and error and error type generally are.